### PR TITLE
cgit: update dependencies

### PR DESCRIPTION
**Pull request by iefdev**
#### Description

In the latest version of CGit they have added support for: `lz, xz, zstd` [»»»](https://git.zx2c4.com/cgit/log/)

By replacing `zlib` and `libiconv` with `libarchive` and `lzip`… CGit will get full support for all the new archive types, incl the replaced dependencies. It'll also make sure `zx` and `zstd` is not on the system as a runtime– &/or extract dependency through another port.

```bash
$ grep Library <(port deps libarchive)
Library Dependencies: bzip2, zlib, libxml2, xz, lzo2, libiconv, lz4, zstd
```

###### Type(s)

- [x] enhancement

###### Tested on

macOS 10.7.5 11G63
Xcode 4.6.3 4H1503

###### Verification

Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vsd install`?

###### Notes

I bumped the revision, but perhaps it's not needed?

### CODE REVIEW

**Comment by ryandesign** [comment]
Thanks!

> I bumped the revision, but perhaps it's not needed?

It's needed, since you've changed the dependencies. If it's helpful, I wrote up some more [guidance on increasing revisions](https://lists.macports.org/pipermail/macports-dev/2020-June/042039.html).

**Comment by ryandesign** [comment]
You probably need to do more than just change the dependencies. You probably also need to also inform the build system that it should use them, and not what it was using before.

Looking in the build log from above, I see that it is still linking with libiconv and zlib. Look at the end of this line:

`/usr/bin/clang -arch x86_64   -g -O2 -Wall -I. -DPRECOMPOSE_UNICODE -DPROTECT_HFS_DEFAULT=1 -I/usr/local/include -DGIT_HOST_CPU="\"x86_64\"" -I/opt/local/include -DNO_CURL -DNO_OPENSSL -DUSE_ST_TIMESPEC -DNO_GETTEXT -DSHA1_APPLE -DSHA256_BLK -DSHA1_MAX_BLOCK_SIZE="1024L*1024L*1024L"  -DHAVE_DEV_TTY -DHAVE_BSD_SYSCTL -DHAVE_GETDELIM -DHAVE_NS_GET_EXECUTABLE_PATH  -DAPPLE_COMMON_CRYPTO -DFREAD_READS_DIRECTORIES -DCOMMON_DIGEST_FOR_OPENSSL -DNO_MEMMEM -Icompat/regex -DSHELL_PATH='"/bin/sh"' -DPAGER_ENV='"LESS=FRX LV=-c"' -o ../cgit   -L/usr/local/lib -L/opt/local/lib ../memrchr.o ../cgit.o ../cache.o ../cmd.o ../configfile.o ../filter.o ../html.o ../parsing.o ../scan-tree.o ../shared.o ../ui-atom.o ../ui-blame.o ../ui-blob.o ../ui-clone.o ../ui-commit.o ../ui-diff.o ../ui-log.o ../ui-patch.o ../ui-plain.o ../ui-refs.o ../ui-repolist.o ../ui-shared.o ../ui-snapshot.o ../ui-ssdiff.o ../ui-stats.o ../ui-summary.o ../ui-tag.o ../ui-tree.o common-main.o libgit.a xdiff/lib.a  -lz  -liconv`

The `-liconv` flag is coming from the portfile itself:
```
configure.ldflags-append -liconv
```
I don't see `-lz` in the portfile so that must be coming from the build system.

(macOS ships with both zlib and libiconv so the port is now linking with those if the MacPorts ones don't exist, and that's probably not what we want.)

Unrelated to your changes, I also see `-L/usr/local/lib` in the compile line, which should be removed. We don't want to find libraries there.

**Comment by iEFdev** [comment]
> (macOS ships with both zlib and libiconv so the port is now linking with those if the MacPorts ones don't exist, and that's probably not what we want.)

They do exist. That was the nice thing about adding `libarchive`, that `zlib` and `libiconv` comes in that way + the new ones. And the log reports they are in the tree.

I found a `LIB` in `cgit.mk`, that produced:
```
# $(info    LIBS:            $(LIBS))
LIBS:            libgit.a xdiff/lib.a  -lz  -liconv
```
…after the Makefile was included.

—

Yes, I did see that (`/usr/local …`) in my log to now. I've located it now. It's in a file from git that is moved in there in post-extract.

I was thinking about patching the (cgit) Makefile to not include it, and add the section there. There's nothing more in that file we need, but in case other files in that git dir… I patched the file itself.

Just tested a reinstall, and no more `-I/usr/local …`/`-L/usr/local …`. 🤞

**Comment by ryandesign** [comment]
If it's your intention that cgit will continue to link directly with MacPorts zlib and libiconv, then it should continue to declare those dependencies.

**Comment by iEFdev** [comment]
Ok, even if `libarchive` have both in its list of dependencies?

**Comment by ryandesign** [comment]
Yes. What libarchive depends on is irrelevant. If cgit directly links with zlib (i.e. it compiles with the `-lz` flag), it should list zlib in depends_lib. Same for libiconv and any other libraries it directly links with.

**Comment by iEFdev** [comment]
Ah, ok. Thanks! I thought (hoped) it would link it anyway because of the dep. chain.

**Comment by ryandesign** [comment]
It will work your way, but that will make it harder for other maintainers in the future. If zlib or libiconv are updated to a new major library version in the future (granted this is unlikely with those libraries but it could happen), then we need to be able to identify all the ports that link with those libraries so that we can revbump those ports. The way that we would identify those ports is by looking at the dependencies declared in the portfile.

**Comment by ryandesign** [comment]
Also, what if libarchive is changed in the future and no longer depends on zlib or libiconv? That's probably unlikely as well, but if it happens, then you're left with cgit still linking with MacPorts libraries although they are no longer present in the dependency chain, or you're left with cgit now using macOS versions of those libraries.

File: www/cgit/Portfile [hunk 1]
```diff
@@ -88,7 +91,11 @@ livecheck.type  regex
...
-  » ${prefix}/etc/cgitrc.sample
-"
+post-activate {
+    ui_msg "
+::
+:: A sample config file have been placed in:
+::   ${prefix}/etc/cgitrc.sample
+::
+    "
+}
```
Comment (ryandesign) [inline www/cgit/Portfile:101]: Please revert this change. `notes` is the correct way to display post-installation notes to the user.

File: www/cgit/files/patch-config-make-uname.diff [hunk 2]
```diff
@@ -0,0 +1,13 @@
+--- ./git/config.mak.uname.orig	2020-08-18 05:34:19.000000000 +0200
++++ ./git/config.mak.uname	2020-08-18 05:43:50.000000000 +0200
+@@ -133,8 +133,8 @@
+ 	HAVE_BSD_SYSCTL = YesPlease
+ 	FREAD_READS_DIRECTORIES = UnfortunatelyYes
+ 	HAVE_NS_GET_EXECUTABLE_PATH = YesPlease
+-	BASIC_CFLAGS += -I/usr/local/include
+-	BASIC_LDFLAGS += -L/usr/local/lib
++	BASIC_CFLAGS += -I/opt/local/include
++	BASIC_LDFLAGS += -L/opt/local/lib
```
Comment (ryandesign) [inline www/cgit/files/patch-config-make-uname.diff:10]: You can't assume that the user has installed MacPorts into /opt/local.

Comment (ryandesign) [reply www/cgit/files/patch-config-make-uname.diff:10]: MacPorts already adds the right `-I` and `-L` flags for whatever the user's MacPorts prefix is, so you can probably just remove the /usr/local lines here without replacing them with anything.

Comment (iEFdev) [reply www/cgit/files/patch-config-make-uname.diff:10]: yes, I was thinking about that after I made the diff. It would be enough just to drop those lines.

**Comment by iEFdev** [comment]
Looks like Azure is having some time issues today. Earlier it was about a file was from the “future”, and now an ntp err.

```
0.us.pool.ntp.org 68.233.45.146 no UCST response after 5 seconds
##[error]Bash exited with code '1'.
```

**Comment by ryandesign** [comment]
Explicit time synchronization was added to our CI script specifically in the hopes of fixing the "portfile is from the future" errors. See https://github.com/macports/macports-ports/pull/8138.

**Comment by iEFdev** [comment]
Is this one good to go? I dropped the 2 lines and reverted the change to notes.

**Comment by ryandesign** [comment]
I installed the port with your changes and the only binary the port installs is only linked with zlib and libiconv:

```
$ otool -L /opt/local/www/data/cgit/cgit.cgi
/opt/local/www/data/cgit/cgit.cgi:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/local/lib/libiconv.2.dylib (compatibility version 9.0.0, current version 9.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

I think that the only way it uses xz, bzip2, zip, zstd, lzip is that it calls their executables at runtime, so those should probably be in depends_run.

How did you determine that it uses libarchive? I couldn't find anything about that.

I'm also not clear how or if the port uses git or openssl, or whether they should be moved to depends_run as well. Those dependencies have been in the port since it was added 13 years ago.

**Comment by iefdev** [comment]
> How did you determine that it uses libarchive? I couldn't find anything about that.

That part was to bring the its dependencies into cgit, to make sure they're installed. libarchive contains all the needed ones. 

```
$  port deps libarchive
Full Name: libarchive @3.4.3_0
Build Dependencies:   pkgconfig
Library Dependencies: bzip2, zlib, libxml2, xz, lzo2, libiconv, lz4, zstd
```
_I also think libarchive brings a newer version of tar, if I'm not mistaken._

Do you want me to specify them (each) manually instead? …and as depends_run?

—

The git, and openssl deps… I haven't reflected over those, since they're were there before, but maybe it could (or even should) be moved to depends_run.

The port downloads its own git version to pick some files before it builds, so I guess that (the port) git isn't a “build” dep.

Probably the depens_build should only have libiconv and zlib, and the other are depends_run.

**Comment by iefdev** [comment]
I noticed the modeline was bad, so I updated it and repushed.

**Comment by g5pw** [comment]
This looks ok to me, is it ready to be merged?

**Comment by kencu** [comment]
why does it need openssl to run, but not to build? That's weird.
